src/v/redpanda/admin_server.cc

```diff
@@ -349,19 +349,89 @@ ss::future<ss::httpd::redirect_exception> admin_server::redirect_to_leader(
             leader_id_opt.value()),
           ss::httpd::reply::status_type::service_unavailable);
     }
+    auto leader = leader_opt.value();
+
+    // Heuristic for finding peer's admin API interface that is accessible
+    // from the client that sent this request:
+    // - if the host in the Host header matches one of our advertised kafka
+    //   addresses, then assume that the peer's advertised kafka address
+    //   with the same index will also be their public admin API address.
+    // - Assume that the peer is listening on the same port that the client
+    //   used to make this request (i.e. the port in Host)
+    //
+    // This will work reliably if all node configs have symmetric kafka listener
+    // sections (i.e. all specify the same number of listeners in the same
+    // order, for example all nodes have an internal and an external listener in
+    // that order), and the hostname used for connecting to the admin API
+    // matches one of the hostnames used for a kafka listener.
+    //
+    // The generic fallback if the heuristic fails is to use the peer's
+    // internal RPC address.  This works if the user is e.g. connecting
+    // by IP address to a k8s cluster's internal pod IP.
```

Comment:
🙇

Reply:
Is there a safety mechanism to avoid infinitely redirecting admin requests 🤔 ?

I'm thinking of a corner case where the configs in each node will make this heuristic redirect requests between a pair of nodes that are not leaders. Do you think would be possible?

Reply:
oo good question!

Reply:
The primary safety mechanism for infinite redirects is client side -- a library like `requests` will impose a maximum number of redirects that they'll follow before throwing an error (same for browsers etc).

We're also protected by the way the controller leader is determined in redpanda, in that if a node has a wrong idea of the leadership, there is a time bound (the voting timeout) on how long it will continue to hold that view if the alleged leader isn't really sending heartbeats regularly.  It follows that two nodes cannot both think the other is the leader for more than that timeout.

We could be tricked into a redirect loop by a malicious node configuration, if the leader advertised a kafka address that actually went to a different node, but in that case the client side protection would kick in.

Reply:
WDYT about adding a special header to limit the max amount of redirects?
Something like `X-REDPANDA-ADMIN-REDIRECT-COUNT`, increase it in each redirect and stop once it reaches a predefined max redirect which I guess can be something small like 3 or 5?

Reply:
Hmm, it's pretty normal for HTTP servers to issue redirects without checking for cycles -- the HTTP spec includes that as a "SHOULD" for clients (https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3).  I'm not sure I see the value of doing an additional check server side.

I just double checked that the standard go http client (i.e. cloud, rpk) implements the redirect limit -- it does, with a threshold of 10 redirects.

Reply:
Oh ok! so Redpanda is simply returning a 301?
I thought it was proxying the request internally.

Reply:
A 307, but yes -- we're just asking the client to talk to another node. That's why the address-guessing is important -- we know how to talk to our peer, but we have to guess how the client can talk to them.